src/main/scala/org/economicsl/auctions/sandbox.sc

```diff
@@ -53,4 +53,24 @@ val orderBook5 = orderBook4 + order8
 
 // take a look at paired orders
 val (pairedOrders, _) = orderBook5.takeWhileMatched
-pairedOrders.toList
\ No newline at end of file
+pairedOrders.toList
+
+
+// Why can this not be made covariant in T? Can divisible orders be constructed using an auxiliary constructor on LimitAskOrder?
+case class DivisibleLimitAskOrder[T <: Tradable](issuer: UUID, limit: Price, quantity: Quantity, tradable: T) extends
+  LimitAskOrder[T] with Divisible[T, DivisibleLimitAskOrder[T]] {
+
+  def split(residual: Quantity): (DivisibleLimitAskOrder[T], DivisibleLimitAskOrder[T]) = {
+    require(residual.value < this.quantity.value)  // can this be check be lifted into the type system?
```

Comment:
@davidrpugh I'm not sure if this test can be lifted into the type system. Because for that to work, the values would have to be known at compile time, wouldn't they? That is at least how it works in C++. In C++, templates can, for example, have value parameters: 

```
template <int param> Foo {...}
```
You can then perform compile-time checks, e.g. 

```
static_assert(param < 5);
```

Alternatively, you can also define another template that performs the check by only providing a specialisation for the 'true' case: 

```
template <class T> struct check; 
template<> struct check<true> {};
```

In order to perform the check at compile time, `Foo` could then inherit from `check` which would produce a compiler error if the condition fails (because there is no specialisation for 'false'): 

```
template <int param> Foo : check<(param < 5)> { ... } // fails at compile time if param >= 5
```

The obvious problem with the whole approach is that the template parameters need to be known at compile time. It would be interesting to know if there is any other solution in Scala.